feline/src/main/java/com/spotify/feline/FelineTransformer.java

```diff
@@ -65,24 +68,47 @@ static boolean onEnter(
 
       boolean state = FelineRuntime.STATE.get();
 
-      if (!future.isDone() && !state) {
-        FelineRuntime.accept(typeName + "." + methodName + methodSig);
+      if (state) {
+        // Already inside a blocking operation, fast-exit
+        // and keep state as true (i.e. do nothing)
+        return Collections.emptyMap();
+      }
 
-        // Set state to true as to ignore any nested blocked calls, e.g. where one Future delegates
-        // to another.
-        // The state must be set after consumers are invoked above. Consumers can throw exceptions,
-        // in which case we
-        // must not have modified the state first as it would not be reset in onExit().
-        FelineRuntime.STATE.set(true);
+      if (future.isDone()) {
+        // This method call is not actually blocking, so fast-exit
+        // and keep state as false (i.e. do nothing)
+        return Collections.emptyMap();
       }
 
-      return state;
+      final String methodCall = typeName + "." + methodName + methodSig;
+      FelineRuntime.accept(methodCall);
+
+      // Set state to true as to ignore any nested blocked calls, e.g. where one Future delegates
+      // to another.
+      // The state must be set after consumers are invoked above. Consumers can throw exceptions,
+      // in which case we
+      // must not have modified the state first as it would not be reset in onExit().
+      FelineRuntime.STATE.set(true);
+
+      // Can't use custom classes here, since they won't be visible to classes in standard library
+      // such as Future.
+      final Map<String, Object> data = new HashMap<>(4);
```

Comment:
could this be an initial size of `2`? the map seems to only ever contain two entries, and a size of 2 would still avoid any resizing (until someone adds a 3rd item to the map one day)

Reply:
Yes, I think I originally had 3 items in the map but over the course of refactoring it got reduced to 2.
I also considered using a plain Object[], but it would get awkward for consumers to use.

I can change it back to 2.